File: WinForms/DevExpress.OutlookInspiredApp.Win/CS/DevExpress.OutlookInspiredApp.Win/ViewModels/FilterTreeViewModel.cs
namespace DevExpress.DevAV.ViewModels {
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Collections.ObjectModel;
    using System.ComponentModel.DataAnnotations;
    using System.Linq.Expressions;
    using DevExpress.Data.Filtering;
    using DevExpress.Data.Filtering.Helpers;
    using DevExpress.Data.Utils;
    using DevExpress.DevAV.Common.ViewModel;
    using DevExpress.Mvvm;
    using DevExpress.Mvvm.DataAnnotations;
    using DevExpress.Mvvm.DataModel;
    using DevExpress.Mvvm.POCO;

    public class FilterTreeViewModel<TEntity, TPrimaryKey, TUnitOfWork> : FilterTreeViewModelBase
        where TEntity : class
        where TUnitOfWork : class, IUnitOfWork {
        public const string StaticFiltersName = "Favorites";
        public const string CustomFiltersName = "Custom Filters";
        public const string GroupFiltersName = "Groups";

        public FilterTreeViewModel(
            CollectionViewModel<TEntity, TPrimaryKey, TUnitOfWork> colViewModel,
            IFilterTreeModelPageSpecificSettings settings) :
            base(settings) {
            SetViewModel(colViewModel);
            ViewModelHelper.EnsureViewModel(this, colViewModel);
            Init();
            ISupportCustomFilters scf = colViewModel as ISupportCustomFilters;
            if(scf != null)
                scf.CustomFiltersReset += scf_CustomFiltersReset;
        }
        void scf_CustomFiltersReset(object sender, EventArgs e) {
            ResetCustomFilters();
        }
        protected internal CollectionViewModel<TEntity, TPrimaryKey, TUnitOfWork> CollectionViewModel {
            get { return (CollectionViewModel<TEntity, TPrimaryKey, TUnitOfWork>)ViewModel; }
        }
        protected virtual void OnSelectedItemChanged() {
            this.RaiseCanExecuteChanged(x => x.Select(null));
            RaiseSelectedItemChanged();
            ApplyFilter(SelectedItem);
        }
        public bool CanSelectModule(FilterItem item) {
            return SelectedItem != item;
        }
        [Command]
        public void Select(FilterItem item) {
            SelectedItem = item;
        }
        public event EventHandler SelectedItemChanged;
        void RaiseSelectedItemChanged() {
            if(SelectedItemChanged != null)
                SelectedItemChanged(this, EventArgs.Empty);
        }
        void ApplyFilter(FilterItemBase fItem) {
            if(fItem != null && !object.ReferenceEquals(fItem.FilterCriteria, null))
                CollectionViewModel.FilterExpression = GetFilterExpression(fItem);
            else
                CollectionViewModel.FilterExpression = null;
        }
        public event EventHandler FilterAccordionChanged;
        void RaiseFilterAccordionChanged() {
            if(FilterAccordionChanged != null)
                FilterAccordionChanged(this, EventArgs.Empty);
        }
        protected override FilterItemBase CreateFilterItem(string name, CriteriaOperator filterCriteria, string imageUri) {
            return FilterItem.Create(name, filterCriteria);
        }
        
        [Command]
        public void New() {
            var newFilterItem = CreateFilterItem(null, null, null);
            CustomFilterViewModel customViewModel = CreateCustomFilterViewModel<CustomFilterViewModel>(newFilterItem);
            if(ShowFilterDialog(customViewModel, "Custom Filter")) {
                if(customViewModel.Save) {
                    AddNewCustomFilter(newFilterItem);
                    RaiseFilterAccordionChanged();
                }
                SelectedItem = newFilterItem;
            }
        }
        [Command]
        public void Modify(FilterItem item) {
            CustomFilterViewModel customViewModel = CreateCustomFilterViewModel<CustomFilterViewModel>(item);
            if(ShowFilterDialog(customViewModel, "Custom Filter")) {
                if(customViewModel.Save) {
                    SaveCustomFilters();
                    RaiseFilterAccordionChanged();
                }
                ApplyFilter(item);
            }
        }
        [Command]
        public void Delete(FilterItem item) {
            DeleteCustomFilter(item);
            RaiseFilterAccordionChanged();
            if(SelectedItem == item)
                SelectedItem = null;
        }
        [Command]
        public void NewGroup() {
            NewGroupCore(CreateFilterItem(null, null, null));
        }
        [Command]
        public void NewGroupFromSelection(IEnumerable<TEntity> selection) {
            NewGroupCore(CreateFilterItem(null, CollectionViewModel.GetInOperator(selection), null));
        }
        public virtual ObservableCollection<FilterItemBase> Groups {
            get;
            protected set;
        }
        public virtual void AddNewGroupFilter(FilterItemBase fItem) {
            Groups.Add(fItem);
            SaveGroupFilters();
        }
        public virtual void DeleteGroupFilter(FilterItemBase fItem) {
            Groups.Remove(fItem);
            SaveGroupFilters();
        }
        public virtual void ModifyGroupFilter(FilterItemBase fItem) {
            SaveGroupFilters();
        }
        void SaveGroupFilters() {
            settingsCore.GroupFilters = SaveToSettings(Groups);
            settingsCore.Settings.Save();
        }
        public override void Init() {
            Groups = CreateFilterItems(settingsCore.GroupFilters);
            base.Init();
        }
        void NewGroupCore(FilterItemBase newFilterItem) {
            GroupFilterViewModel gViewModel = CreateCustomFilterViewModel<GroupFilterViewModel>(newFilterItem);
            if(ShowFilterDialog(gViewModel, "Group Filter")) {
                if(gViewModel.Save) {
                    AddNewGroupFilter(newFilterItem);
                    RaiseFilterAccordionChanged();
                }
                SelectedItem = newFilterItem;
            }
        }
        [Command]
        public void ModifyGroup(FilterItem item) {
            GroupFilterViewModel gViewModel = CreateCustomFilterViewModel<GroupFilterViewModel>(item);
            if(ShowFilterDialog(gViewModel, "Group Filter")) {
                if(gViewModel.Save) {
                    ModifyGroupFilter(item);
                    RaiseFilterAccordionChanged();
                }
                ApplyFilter(item);
            }
        }
        [Command]
        public void DeleteGroup(FilterItem item) {
            DeleteGroupFilter(item);
            RaiseFilterAccordionChanged();
            if(SelectedItem == item)
                SelectedItem = null;
        }
        public override void ResetCustomFilters() {
            base.ResetCustomFilters();
            RaiseFilterAccordionChanged();
        }
        protected virtual T CreateCustomFilterViewModel<T>(FilterItemBase fItem) where T : FilterViewModelBase, new() {
            T filterViewModel = ViewModelSource.Create<T>();
            ViewModelHelper.EnsureViewModel(filterViewModel, CollectionViewModel, fItem);
            return filterViewModel;
        }
        bool ShowFilterDialog(FilterViewModelBase filterViewModel, string key) {
            var service = this.GetService<IDocumentManagerService>(key);
            if(service != null) {
                var document = service.CreateDocument(key, filterViewModel, null, CollectionViewModel);
                filterViewModel.Document = document;
                document.Show();
                return filterViewModel.Result.GetValueOrDefault();
            }
            return false;
        }
        protected virtual bool EnableGroups {
            get { return true; }
        }
        public string GetFilterName(object filtersCollection, FilterItemBase filter) {
            if(filter != null) {
                var count = CollectionViewModel.GetEntitiesCount(GetFilterExpression(filter));
                if(count > 0)
                    return filter.Name + " (" + count + ")";
                else
                    return filter.Name;
            }
            else {
                if(object.Equals(filtersCollection, StaticFilters))
                    return StaticFiltersName;
                if(object.Equals(filtersCollection, CustomFilters))
                    return CustomFiltersName;
                if(object.Equals(filtersCollection, Groups))
                    return GroupFiltersName;
            }
            return null;
        }
        public IList GetChildren(object dataItem) {
            if(dataItem == this) {
                if(EnableGroups)
                    return (IList)(new List<object> { StaticFilters, CustomFilters, Groups });
                else
                    return (IList)(new List<object> { StaticFilters, CustomFilters });
            }
            if(dataItem is System.Collections.ObjectModel.ObservableCollection<FilterItemBase>)
                return (IList)dataItem;
            return null;
        }
        internal static Expression<Func<TEntity, bool>> GetFilterExpression(CriteriaOperator criteria) {
            return FilterExpressionFactory<TEntity>.CreateFilterExpression(criteria);
        }
        static Expression<Func<TEntity, bool>> GetFilterExpression(FilterItemBase filter) {
            return FilterExpressionFactory<TEntity>.CreateFilterExpression(filter.FilterCriteria);
        }
        #region Filter Item ViewModels
        public class FilterItem : FilterItemBase {
            public static FilterItem Create(string filterName, CriteriaOperator filterCriteria) {
                return ViewModelSource.Create(() => new FilterItem(filterName, filterCriteria));
            }
            protected FilterItem(string name, CriteriaOperator filterCriteria) {
                this.Name = name;
                this.FilterCriteria = filterCriteria;
            }
        }
        #endregion Items
    }
    #region Custom Filter ViewModel
    [POCOViewModel(ImplementIDataErrorInfo = true)]
    public abstract class FilterViewModelBase : ISupportParameter {
        public FilterViewModelBase() {
            Save = true;
        }
        FilterItemBase filterItem;
        public IDocument Document {
            get;
            set;
        }
        public bool? Result {
            get;
            private set;
        }
        public virtual bool Save {
            get;
            set;
        }
        [Required]
        public virtual string Name {
            get;
            set;
        }
        public CriteriaOperator FilterCriteria {
            get { return filterItem.FilterCriteria; }
        }
        public event EventHandler<QueryFilterCriteriaEventArgs> QueryFilterCriteria;
        void RaiseQueryFilterCriteria() {
            EventHandler<QueryFilterCriteriaEventArgs> handler = QueryFilterCriteria;
            if(handler != null)
                handler(this, new QueryFilterCriteriaEventArgs(filterItem));
        }
        protected IMessageBoxService MessageBoxService {
            get { return this.GetService<IMessageBoxService>(); }
        }
        protected abstract string GetDefaultName();
        
        [Command]
        public void OK() {
            Result = true;
            if(string.IsNullOrEmpty(Name))
                Name = GetDefaultName();
            if(Save)
                filterItem.Name = Name;
            RaiseQueryFilterCriteria();
            Document.Close();
        }
        [Command]
        public void Cancel() {
            Result = false;
            Document.Close();
        }
        object ISupportParameter.Parameter {
            get { return filterItem; }
            set {
                filterItem = (FilterItemBase)value;
                Name = filterItem.Name;
            }
        }
    }
    public class QueryFilterCriteriaEventArgs : EventArgs {
        FilterItemBase item;
        public QueryFilterCriteriaEventArgs(FilterItemBase item) {
            this.item = item;
        }
        public CriteriaOperator FilterCriteria {
            get { return item.FilterCriteria; }
            set { item.FilterCriteria = value; }
        }
    }
    public class CustomFilterViewModel : FilterViewModelBase {
        static int id = 0;
        protected override string GetDefaultName() {
            return "Custom Filter " + (id++).ToString();
        }
    }
    public class GroupFilterViewModel : FilterViewModelBase {
        static int id = 0;
        protected override string GetDefaultName() {
            return "Group " + (id++).ToString();
        }
    }
    #endregion
    
    static class FilterExpressionFactory<TEntity> {
        public static Expression<Func<TEntity, bool>> CreateFilterExpression(CriteriaOperator criteria) {
            try {
                var nullValueSafeCriteria = criteria.ReferenceEqualsNull() ?
                    criteria : criteria.Accept(new NullValueToNullString());
                var caseInsensitiveCriteria = Data.Helpers.StringsTolowerCloningHelper.Process(nullValueSafeCriteria);
                return CriteriaOperatorToExpressionConverter.GetGenericWhere<TEntity>(caseInsensitiveCriteria);
            }
            catch(Exception e) {
                throw new NotSupportedException("Error in Filter:" + CriteriaOperator.ToString(criteria), e);
            }
        }
        sealed class NullString : ICustomFunctionOperator, ICustomFunctionOperatorConvertibleToExpression {
            internal static readonly NullString Instance = new NullString();
            public string Name { get { return "NullString"; } }
            public Expression Convert(Data.Linq.ICriteriaToExpressionConverter converter, params Expression[] operands) {
                return Expression.Constant(null, typeof(string));
            }
            public object Evaluate(params object[] operands) {
                return null;
            }
            public Type ResultType(params Type[] operands) {
                return typeof(string);
            }
        }
        sealed class NullValueToNullString : Data.Filtering.Helpers.ClientCriteriaClonerBase.DeepValuesCloneBase {
            static NullValueToNullString() {
                CriteriaOperator.RegisterCustomFunction(NullString.Instance);
            }
            public override CriteriaOperator Visit(FunctionOperator theOperator) {
                if(theOperator.OperatorType >= FunctionOperatorType.StartsWith && theOperator.OperatorType <= FunctionOperatorType.EndsWith) {
                    OperandValue value = null;
                    if(theOperator.Operands[1].Is(out value) && value.XmlValue is NullValue)
                        theOperator.Operands[1] = new FunctionOperator(FunctionOperatorType.Custom, "NullString");
                }
                return base.Visit(theOperator);
            }
        }
    }
}

File: WinForms/DevExpress.OutlookInspiredApp.Win/VB/DevExpress.OutlookInspiredApp.Win/ViewModels/FilterTreeViewModel.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.ComponentModel.DataAnnotations
Imports System.Linq.Expressions
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Filtering.Helpers
Imports DevExpress.Data.Utils
Imports DevExpress.DevAV.Common.ViewModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.DataModel
Imports DevExpress.Mvvm.POCO

Namespace DevExpress.DevAV.ViewModels

    Public Class FilterTreeViewModel(Of TEntity As Class, TPrimaryKey, TUnitOfWork As {Class, IUnitOfWork})
        Inherits FilterTreeViewModelBase

        Public Const StaticFiltersName As String = "Favorites"

        Public Const CustomFiltersName As String = "Custom Filters"

        Public Const GroupFiltersName As String = "Groups"

        Public Sub New(ByVal colViewModel As CollectionViewModel(Of TEntity, TPrimaryKey, TUnitOfWork), ByVal settings As IFilterTreeModelPageSpecificSettings)
            MyBase.New(settings)
            SetViewModel(colViewModel)
            EnsureViewModel(Me, colViewModel)
            Init()
            Dim scf As ISupportCustomFilters = TryCast(colViewModel, ISupportCustomFilters)
            If scf IsNot Nothing Then AddHandler scf.CustomFiltersReset, AddressOf scf_CustomFiltersReset
        End Sub

        Private Sub scf_CustomFiltersReset(ByVal sender As Object, ByVal e As EventArgs)
            ResetCustomFilters()
        End Sub

        Protected Friend ReadOnly Property CollectionViewModel As CollectionViewModel(Of TEntity, TPrimaryKey, TUnitOfWork)
            Get
                Return CType(ViewModel, CollectionViewModel(Of TEntity, TPrimaryKey, TUnitOfWork))
            End Get
        End Property

        Protected Overridable Sub OnSelectedItemChanged()
            RaiseCanExecuteChanged(Sub(x) x.Select(Nothing))
            RaiseSelectedItemChanged()
            ApplyFilter(SelectedItem)
        End Sub

        Public Function CanSelectModule(ByVal item As FilterItem) As Boolean
            Return SelectedItem IsNot item
        End Function

        <Command>
        Public Sub [Select](ByVal item As FilterItem)
            SelectedItem = item
        End Sub

        Public Event SelectedItemChanged As EventHandler

        Private Sub RaiseSelectedItemChanged()
            RaiseEvent SelectedItemChanged(Me, EventArgs.Empty)
        End Sub

        Private Sub ApplyFilter(ByVal fItem As FilterItemBase)
            If fItem IsNot Nothing AndAlso Not ReferenceEquals(fItem.FilterCriteria, Nothing) Then
                CollectionViewModel.FilterExpression = GetFilterExpression(fItem)
            Else
                CollectionViewModel.FilterExpression = Nothing
            End If
        End Sub

        Public Event FilterAccordionChanged As EventHandler

        Private Sub RaiseFilterAccordionChanged()
            RaiseEvent FilterAccordionChanged(Me, EventArgs.Empty)
        End Sub

        Protected Overrides Function CreateFilterItem(ByVal name As String, ByVal filterCriteria As CriteriaOperator, ByVal imageUri As String) As FilterItemBase
            Return FilterItem.Create(name, filterCriteria)
        End Function

        <Command>
        Public Sub [New]()
            Dim newFilterItem = CreateFilterItem(Nothing, Nothing, Nothing)
            Dim customViewModel As CustomFilterViewModel = CreateCustomFilterViewModel(Of CustomFilterViewModel)(newFilterItem)
            If ShowFilterDialog(customViewModel, "Custom Filter") Then
                If customViewModel.Save Then
                    AddNewCustomFilter(newFilterItem)
                    RaiseFilterAccordionChanged()
                End If

                SelectedItem = newFilterItem
            End If
        End Sub

        <Command>
        Public Sub Modify(ByVal item As FilterItem)
            Dim customViewModel As CustomFilterViewModel = CreateCustomFilterViewModel(Of CustomFilterViewModel)(item)
            If ShowFilterDialog(customViewModel, "Custom Filter") Then
                If customViewModel.Save Then
                    SaveCustomFilters()
                    RaiseFilterAccordionChanged()
                End If

                ApplyFilter(item)
            End If
        End Sub

        <Command>
        Public Sub Delete(ByVal item As FilterItem)
            DeleteCustomFilter(item)
            RaiseFilterAccordionChanged()
            If SelectedItem Is item Then SelectedItem = Nothing
        End Sub

        <Command>
        Public Sub NewGroup()
            NewGroupCore(CreateFilterItem(Nothing, Nothing, Nothing))
        End Sub

        <Command>
        Public Sub NewGroupFromSelection(ByVal selection As IEnumerable(Of TEntity))
            NewGroupCore(CreateFilterItem(Nothing, CollectionViewModel.GetInOperator(selection), Nothing))
        End Sub

        Public Overridable Property Groups As ObservableCollection(Of FilterItemBase)

        Public Overridable Sub AddNewGroupFilter(ByVal fItem As FilterItemBase)
            Groups.Add(fItem)
            SaveGroupFilters()
        End Sub

        Public Overridable Sub DeleteGroupFilter(ByVal fItem As FilterItemBase)
            Groups.Remove(fItem)
            SaveGroupFilters()
        End Sub

        Public Overridable Sub ModifyGroupFilter(ByVal fItem As FilterItemBase)
            SaveGroupFilters()
        End Sub

        Private Sub SaveGroupFilters()
            settingsCore.GroupFilters = SaveToSettings(Groups)
            settingsCore.Settings.Save()
        End Sub

        Public Overrides Sub Init()
            Groups = CreateFilterItems(settingsCore.GroupFilters)
            MyBase.Init()
        End Sub

        Private Sub NewGroupCore(ByVal newFilterItem As FilterItemBase)
            Dim gViewModel As GroupFilterViewModel = CreateCustomFilterViewModel(Of GroupFilterViewModel)(newFilterItem)
            If ShowFilterDialog(gViewModel, "Group Filter") Then
                If gViewModel.Save Then
                    AddNewGroupFilter(newFilterItem)
                    RaiseFilterAccordionChanged()
                End If

                SelectedItem = newFilterItem
            End If
        End Sub

        <Command>
        Public Sub ModifyGroup(ByVal item As FilterItem)
            Dim gViewModel As GroupFilterViewModel = CreateCustomFilterViewModel(Of GroupFilterViewModel)(item)
            If ShowFilterDialog(gViewModel, "Group Filter") Then
                If gViewModel.Save Then
                    ModifyGroupFilter(item)
                    RaiseFilterAccordionChanged()
                End If

                ApplyFilter(item)
            End If
        End Sub

        <Command>
        Public Sub DeleteGroup(ByVal item As FilterItem)
            DeleteGroupFilter(item)
            RaiseFilterAccordionChanged()
            If SelectedItem Is item Then SelectedItem = Nothing
        End Sub

        Public Overrides Sub ResetCustomFilters()
            MyBase.ResetCustomFilters()
            RaiseFilterAccordionChanged()
        End Sub

        Protected Overridable Function CreateCustomFilterViewModel(Of T As {FilterViewModelBase, New})(ByVal fItem As FilterItemBase) As T
            Dim filterViewModel As T = ViewModelSource.Create(Of T)()
            EnsureViewModel(filterViewModel, CollectionViewModel, fItem)
            Return filterViewModel
        End Function

        Private Function ShowFilterDialog(ByVal filterViewModel As FilterViewModelBase, ByVal key As String) As Boolean
            Dim service = GetService(Of IDocumentManagerService)(key)
            If service IsNot Nothing Then
                Dim document = service.CreateDocument(key, filterViewModel, Nothing, CollectionViewModel)
                filterViewModel.Document = document
                document.Show()
                Return filterViewModel.Result.GetValueOrDefault()
            End If

            Return False
        End Function

        Protected Overridable ReadOnly Property EnableGroups As Boolean
            Get
                Return True
            End Get
        End Property

        Public Function GetFilterName(ByVal filtersCollection As Object, ByVal filter As FilterItemBase) As String
            If filter IsNot Nothing Then
                Dim count = CollectionViewModel.GetEntitiesCount(GetFilterExpression(filter))
                If count > 0 Then
                    Return filter.Name & " (" & count & ")"
                Else
                    Return filter.Name
                End If
            Else
                If Equals(filtersCollection, StaticFilters) Then Return StaticFiltersName
                If Equals(filtersCollection, CustomFilters) Then Return CustomFiltersName
                If Equals(filtersCollection, Groups) Then Return GroupFiltersName
            End If

            Return Nothing
        End Function

        Public Function GetChildren(ByVal dataItem As Object) As IList
            If dataItem Is Me Then
                If EnableGroups Then
                    Return CType((New List(Of Object) From {StaticFilters, CustomFilters, Groups}), IList)
                Else
                    Return CType((New List(Of Object) From {StaticFilters, CustomFilters}), IList)
                End If
            End If

            If TypeOf dataItem Is ObservableCollection(Of FilterItemBase) Then Return CType(dataItem, IList)
            Return Nothing
        End Function

        Friend Shared Function GetFilterExpression(ByVal criteria As CriteriaOperator) As Expression(Of Func(Of TEntity, Boolean))
            Return FilterExpressionFactory(Of TEntity).CreateFilterExpression(criteria)
        End Function

        Private Shared Function GetFilterExpression(ByVal filter As FilterItemBase) As Expression(Of Func(Of TEntity, Boolean))
            Return FilterExpressionFactory(Of TEntity).CreateFilterExpression(filter.FilterCriteria)
        End Function

#Region "Filter Item ViewModels"
        Public Class FilterItem
            Inherits FilterItemBase

            Public Shared Function Create(ByVal filterName As String, ByVal filterCriteria As CriteriaOperator) As FilterItem
                Return ViewModelSource.Create(Function() New FilterItem(filterName, filterCriteria))
            End Function

            Protected Sub New(ByVal name As String, ByVal filterCriteria As CriteriaOperator)
                Me.Name = name
                Me.FilterCriteria = filterCriteria
            End Sub
        End Class
#End Region  ' Items
    End Class

#Region "Custom Filter ViewModel"
    <POCOViewModel(ImplementIDataErrorInfo:=True)>
    Public MustInherit Class FilterViewModelBase
        Implements ISupportParameter

        Private _Result As Boolean?

        Public Sub New()
            Save = True
        End Sub

        Private filterItem As FilterItemBase

        Public Property Document As IDocument

        Public Property Result As Boolean?
            Get
                Return _Result
            End Get

            Private Set(ByVal value As Boolean?)
                _Result = value
            End Set
        End Property

        Public Overridable Property Save As Boolean

        <Required>
        Public Overridable Property Name As String

        Public ReadOnly Property FilterCriteria As CriteriaOperator
            Get
                Return filterItem.FilterCriteria
            End Get
        End Property

        Public Event QueryFilterCriteria As EventHandler(Of QueryFilterCriteriaEventArgs)

        Private Sub RaiseQueryFilterCriteria()
            Dim handler As EventHandler(Of QueryFilterCriteriaEventArgs) = QueryFilterCriteriaEvent
            If handler IsNot Nothing Then handler(Me, New QueryFilterCriteriaEventArgs(filterItem))
        End Sub

        Protected ReadOnly Property MessageBoxService As IMessageBoxService
            Get
                Return GetService(Of IMessageBoxService)()
            End Get
        End Property

        Protected MustOverride Function GetDefaultName() As String

        <Command>
        Public Sub OK()
            Result = True
            If String.IsNullOrEmpty(Name) Then Name = GetDefaultName()
            If Save Then filterItem.Name = Name
            RaiseQueryFilterCriteria()
            Document.Close()
        End Sub

        <Command>
        Public Sub Cancel()
            Result = False
            Document.Close()
        End Sub

        Private Property Parameter As Object Implements ISupportParameter.Parameter
            Get
                Return filterItem
            End Get

            Set(ByVal value As Object)
                filterItem = CType(value, FilterItemBase)
                Name = filterItem.Name
            End Set
        End Property
    End Class

    Public Class QueryFilterCriteriaEventArgs
        Inherits EventArgs

        Private item As FilterItemBase

        Public Sub New(ByVal item As FilterItemBase)
            Me.item = item
        End Sub

        Public Property FilterCriteria As CriteriaOperator
            Get
                Return item.FilterCriteria
            End Get

            Set(ByVal value As CriteriaOperator)
                item.FilterCriteria = value
            End Set
        End Property
    End Class

    Public Class CustomFilterViewModel
        Inherits FilterViewModelBase

        Private Shared id As Integer = 0

        Protected Overrides Function GetDefaultName() As String
            Return "Custom Filter " & Math.Min(Threading.Interlocked.Increment(id), id - 1).ToString()
        End Function
    End Class

    Public Class GroupFilterViewModel
        Inherits FilterViewModelBase

        Private Shared id As Integer = 0

        Protected Overrides Function GetDefaultName() As String
            Return "Group " & Math.Min(Threading.Interlocked.Increment(id), id - 1).ToString()
        End Function
    End Class

#End Region
    Friend NotInheritable Class FilterExpressionFactory(Of TEntity)

        Public Shared Function CreateFilterExpression(ByVal criteria As CriteriaOperator) As Expression(Of Func(Of TEntity, Boolean))
            Try
                Dim nullValueSafeCriteria = If(criteria.ReferenceEqualsNull(), criteria, criteria.Accept(New NullValueToNullString()))
                Dim caseInsensitiveCriteria = Data.Helpers.StringsTolowerCloningHelper.Process(nullValueSafeCriteria)
                Return CriteriaOperatorToExpressionConverter.GetGenericWhere(Of TEntity)(caseInsensitiveCriteria)
            Catch e As Exception
                Throw New NotSupportedException("Error in Filter:" & CriteriaOperator.ToString(criteria), e)
            End Try
        End Function

        Private NotInheritable Class NullString
            Implements ICustomFunctionOperator, ICustomFunctionOperatorConvertibleToExpression

            Friend Shared ReadOnly Instance As NullString = New NullString()

            Public ReadOnly Property Name As String Implements ICustomFunctionOperator.Name
                Get
                    Return "NullString"
                End Get
            End Property

            Public Function Convert(ByVal converter As Data.Linq.ICriteriaToExpressionConverter, ParamArray operands As Expression()) As Expression Implements ICustomFunctionOperatorConvertibleToExpression.Convert
                Return Expression.Constant(Nothing, GetType(String))
            End Function

            Public Function Evaluate(ParamArray operands As Object()) As Object Implements ICustomFunctionOperator.Evaluate
                Return Nothing
            End Function

            Public Function ResultType(ParamArray operands As Type()) As Type Implements ICustomFunctionOperator.ResultType
                Return GetType(String)
            End Function
        End Class

        Private NotInheritable Class NullValueToNullString
            Inherits Data.Filtering.Helpers.ClientCriteriaClonerBase.DeepValuesCloneBase

            Shared Sub New()
                CriteriaOperator.RegisterCustomFunction(NullString.Instance)
            End Sub

            Public Overrides Function Visit(ByVal theOperator As FunctionOperator) As CriteriaOperator
                If theOperator.OperatorType >= FunctionOperatorType.StartsWith AndAlso theOperator.OperatorType <= FunctionOperatorType.EndsWith Then
                    Dim value As OperandValue = Nothing
                    If theOperator.Operands(1).[Is](value) AndAlso TypeOf value.XmlValue Is NullValue Then theOperator.Operands(1) = New FunctionOperator(FunctionOperatorType.Custom, "NullString")
                End If

                Return MyBase.Visit(theOperator)
            End Function
        End Class
    End Class
End Namespace
